[INFRA-2384] Improve release documentation (#83)
jenkins-infra/release: README.adoc

== Introduction

This repository contains everything needed to release https://github.com/jenkinsci/jenkins[Jenkins core] or some Jenkins components libraries.

This includes:

* Jenkinsfiles for Releasing and Packaging.
* Kubernetes POD templates definition for releasing and packaging environments on Linux and Windows operating systems.
* Any configuration or utility scripts

The release process implies following Git repository:

* https://github.com/jenkinsci/jenkins[jenkinsci/jenkins] or any other git repository containing the Jenkins code to be released.
* https://github.com/jenkinsci/packaging[jenkinsci/packaging] contains tools and configuration to package Jenkins for OS distribution [Debian, Red Hat, SUSE, MSI, WAR]
* https://github.com/jenkins-infra/chart[jenkins-infra/chart] contains Helm chart used to deploy link:release.ci.jenkins.io[release.ci.jenkins.io] on top of Kubernetes.
* https://github.com/jenkins-infra/azure[jenkins-infra/azure] contains the terraform code used to provision infrastructure needed by the release environment

> == Settings

Reviewer: not sure if this is one of the most important categories, should we move it down to the bottom as an appendix ?

Author: I was wondering the same so maybe it makes more sense to move it down

|
| x

| `RELEASE_PROFILE`
| Define a file containing environment variables specific to a release, located in the profile.d directory.
| x
| x

| `SIGN_ALIAS`
| Define code signing certificate name.
| x
| x

| `SIGN_KEYSTORE_FILENAME`
| Define code signing certificate file name.
| x
| x

| `SIGN_KEYSTORE`
| Define signing keystore.
| x
| x

| `SIGN_CERTIFICATE`
| Define code signing certificate file name.
| x
| x

|===

== Process

=== Introduction

The release process is divided in two categories.
The first part that we mention by using the term the *release*, is when we create a new java code release. It relies on the Maven Release Plugin to perform the release. It involves signing with a GPG key and a code signing certificate. At the end of this operation signed maven artifacts are pushed to a Maven repository.

The second part that we name *packaging* is when we retrieve from a Maven repository, the version we want to package. Then we build distribution packages, publish them, promote artifacts between staging and production environment if needed and finally we ensure that our mirrors are up to date.

==== Required

> In order to triggered new release, you must fulfill following requirements:

Reviewer: ```suggestion
In order to trigger a new release, you must fulfil following requirements:
```

. Once triggered, it asks you which release line you want to package for. The release line matches one of the profile defines in https://github.com/jenkins-infra/release/tree/master/profile.d[profile.d], so please carefully review those settings in order to validate that's what you are looking for.

Once the job is done, every package will be published and then mirror synchronized.

==== Validate
Ensure that packages are correctly published on pkg.jenkins.io and correctly signed.

==== Stage
Staging packages is not yet fully supported, more information on link:https://issues.jenkins-ci.org/browse/INFRA-1363[INFRA-1363] and link:https://issues.jenkins-ci.org/browse/INFRA-2608[INFRA-2608]

== Profile
The release profile is used to identify the kind of release we are going to do. They are mainly influenced by following elements:

. Do we want to releases based on different repository branch?
. Do we want to release based on different git repository?

At the moment we identify four release type

. Weekly
. Stable
. Security
. Lts security

[NOTE]
====

> link:https://release.ci.jenkins.io[release.ci.jenkins.io] is designed in a way that we have two generic jobs, one for release and a second one for packaging.
> Then we have one job per release type which triggered the two generic Jenkins jobs with different parameters.
> While it isn't required to trigger a release type job, it just increase visibility and reduce human error factor.

Reviewer 2: I'm not sure where I can find the trigger job.  Could we review that together in one of our training sessions?
```suggestion
link:https://release.ci.jenkins.io[release.ci.jenkins.io] has two generic jobs, one for release and a second one for packaging.
One job per release type triggers the two generic jobs with different parameters.
While it isn't required to trigger a release type job, it increases visibility and reduces the risk of human error.
```

====

=== Weekly

> The weekly release is the default release. It happened automatically every Tuesday as defined by this link:https://github.com/jenkins-infra/release/blob/19685def608c641496e6e2de3d40c275ca5e913d/Jenkinsfile.d/core/weekly#L15[cron].

Reviewer 2: ```suggestion
The weekly release is the default release. It is scheduled every Tuesday as defined by this link:https://github.com/jenkins-infra/release/blob/19685def608c641496e6e2de3d40c275ca5e913d/Jenkinsfile.d/core/weekly#L15[cron].
```

It uses parameters defined in this link:https://github.com/jenkins-infra/release/blob/master/profile.d/weekly[file]

It releases using the repository https://github.com/jenkinsci/jenkins[jenkinsci/jenkins] from the branch master. Artifacts are pushed to the default maven repository 'Releases'.

If for some reason the release job needs to be re-triggered, you can:

. Connect to the Jenkins VPN
. Open your favorite browser to link:https://release.ci.jenkins.io[release.ci.jenkins.io]
. Review the weekly environment https://github.com/jenkins-infra/release/blob/master/profile.d/weekly[file]:
. Trigger the weekly link:https://release.ci.jenkins.io/blue/organizations/jenkins/core%2Fweekly%2Frelease/branches/[job]

[NOTE]
====
You can re-trigger individually the two downstream jobs, release and packaging.

- Re-triggering the release will do a version bump then push new artifacts.
- Re-triggering the packaging job won't published artifacts if they already exist but it will update website html.
====

=== Stable

> A stable release is a manually triggered release that happened around once a month.

Reviewer 2: ```suggestion
A stable release is a manually triggered release that happens around once a month.
Refer to link:https://www.jenkins.io/download/lts/[LTS Release Line] for more detailed information.
```

It uses parameters defined in this link:https://github.com/jenkins-infra/release/blob/master/profile.d/stable[file].

Before triggering a new stable release, some steps are required:

> . Prepare `jenkinsci/jenkins` repository -> missing documentation link.
> . Create a branch on jenkins-infra/release with a branch name that match the release branch from jenkinsci/jenkins like `stable-2.135`
> . Review and update the stable environment https://github.com/jenkins-infra/release/blob/master/profile.d/stable[file] with:
> .. `RELEASE_GIT_BRANCH` set to the `jenkinsci/jenkins` release branch like `stable-2.135`
> .. `JENKINS_VERSION` set to the final release version that will be packaged. If set to 'stable' then the packaging job will try to guess the version based on what was pushed to the maven repository. cfr settings.
> .. `PACKAGING_GIT_BRANCH` set to the appropriated `jenkinsci/packaging` branch
> . Trigger the stable link:https://release.ci.jenkins.io/blue/organizations/jenkins/core%2Fstable%2Frelease/branches/[job]

Reviewer 2: Use 2.235 rather than 2.135 as the example
```suggestion
. Prepare `jenkinsci/jenkins` repository -> missing documentation link.
. Create a branch on jenkins-infra/release with a branch name that match the release branch from jenkinsci/jenkins like `stable-2.235`
. Review and update the stable environment https://github.com/jenkins-infra/release/blob/master/profile.d/stable[file] with:
.. `RELEASE_GIT_BRANCH` set to the `jenkinsci/jenkins` release branch like `stable-2.235`
.. `JENKINS_VERSION` set to the final release version that will be packaged. If set to 'stable' then the packaging job will try to guess the version based on what was pushed to the maven repository. cfr settings.
.. `PACKAGING_GIT_BRANCH` set to the appropriated `jenkinsci/packaging` branch
. Trigger the stable link:https://release.ci.jenkins.io/blue/organizations/jenkins/core%2Fstable%2Frelease/branches/[job]
```

[NOTE]
====
You can re-trigger individually the two downstream jobs, release and packaging.

- Re-triggering the release will do a version bump then push new artifacts.
- Re-triggering the packaging job won't published artifacts if they already exist but it will update website html.
====

=== Security

> The security release follows the same process than the stable one excepted that it also involve promoting git commits from a private repository to the public one and also promoting maven artifacts from a private maven repository to the public one.

Reviewer 2: ```suggestion
The security release follows the same process as the stable one except that it also promotes git commits from a private repository to the public one and also promotes maven artifacts from a private maven repository to the public one.
```
